docs/tutorial/category_theory/category_theory_intro.md

```diff
@@ -0,0 +1,231 @@
+example : X ⟶ Z := g ≫ h
+```
+
+Note in particular the order! The "maps on the right" convention was chosen; `g ≫ h` means "`g` then
+`h`". Type `≫ with `\gg` in VS Code. Here are the theorems which ensure that we have a category.
+
+```lean
+open category_theory.category
+
+example : 𝟙 X ≫ g = g := id_comp C g
+example : g ≫ 𝟙 Y = g := comp_id C g
+example : (f ≫ g) ≫ h = f ≫ (g ≫ h) := assoc C f g h
+example : (f ≫ g) ≫ h = f ≫ g ≫ h := assoc C f g h -- note \gg is right associative
+```
+
+**TODO** : can one not do `𝒞.id_comp g` or anything like that? Is there a better way to write this stuff?
+
+Monomorphisms and epimorphisms are predicates on morphisms and are implemented as typeclasses.
+
+```lean
+variables (f' : W ⟶ X) (h' : Y ⟶ Z)
+
+example [mono g] : f ≫ g = f' ≫ g → f = f' := mono.right_cancellation f f'
+example [epi g] : g ≫ h = g ≫ h' → h = h' := epi.left_cancellation h h'
+```
+
+All six examples above can also be proved with `simp`.
+
+## Getting started with functors
+
+A functor is a map between categories. It is implemented as a structure. The notation for a functor
+from `C` to `D` is `C ⥤ D`. Type `\func` in VS Code for the symbol. Here we demonstrate how to
+evaluate functors on objects and on morphisms, how to show functors preserve the identity morphism
+and composition of morphisms, how to compose functors, and show the notation `𝟭` for the identity
+functor.
+
+
+```lean
+import category_theory.functor
+
+universes v₁ v₂ v₃ u₁ u₂ u₃ -- the order matters
+
+open category_theory
+
+variables (C : Type u₁) [𝒞 : category.{v₁} C]
+variables (D : Type u₂) [𝒟 : category.{v₂} D]
+variables (E : Type u₃) [ℰ : category.{v₃} E]
+
+include 𝒞 𝒟 ℰ
+
+variables (X Y Z : C) (f : X ⟶ Y) (g : Y ⟶ Z)
+
+-- functors
+variables (F : C ⥤ D) (G : D ⥤ E)
+
+example : D := F.obj X -- functor F on objects
+example : F.obj Y ⟶ F.obj Z := F.map g -- functor F on morphisms
+
+-- A functor sends identity objects to identity objects
+example : F.map (𝟙 X) = 𝟙 (F.obj X) := F.map_id X
+
+-- and preserves compositions
+example : F.map (f ≫ g) = (F.map f) ≫ (F.map g) := F.map_comp f g
+
+-- The identity functor is `𝟭`, currently apparently untypesettable in Lean!
+example : C ⥤ C := 𝟭 C
+
+-- The identity functor is definitionally the identity on objects and morphisms
+example : (𝟭 C).obj X = X := rfl -- category_theory.functor.id_obj
+example : (𝟭 C).map f = f := rfl -- category_theory.functor.id_map
+
+-- composition of functors; typeset with `\ggg`; note order
+example : C ⥤ E := F ⋙ G
+
+-- Composition of the identity either way does nothing:
+example : F ⋙ 𝟭 D = F := F.comp_id
+example : 𝟭 C ⋙ F = F := F.id_comp
+
+-- Composition of functors definitionally does the right thing on objects and morphisms
+example : (F ⋙ G).obj X = G.obj (F.obj X) := F.comp_obj G X -- or rfl
+example : (F ⋙ G).map f = G.map (F.map f) := rfl -- or F.comp_map G X Y f
+
+```
+
+One can also check that associativity of composition of functors is definitionally true.
```

Comment:
```suggestion
One can also check that associativity of composition of functors is definitionally true, although we've observed that relying on this can result in slow proofs. (One should rather use the natural isomorphisms provided in `src/category_theory/whiskering.lean`.)
```